Python/pystate.c

```diff
@@ -1831,10 +1831,21 @@ _PyThreadState_DeleteCurrent(PyThreadState *tstate)
     _Py_EnsureTstateNotNULL(tstate);
 #ifdef Py_GIL_DISABLED
     _Py_qsbr_detach(((_PyThreadStateImpl *)tstate)->qsbr);
+    // tstate_delete_common() removes tstate from consideration for
+    // stop-the-worlds. This means that another thread could enable the GIL
+    // before our call to _PyEval_ReleaseLock(), violating its invariant that
+    // the calling thread holds the GIL if and only if the GIL is enabled. Deal
+    // with this by deciding if we need to release the GIL before
+    // tstate_delete_common(), when the invariant is still true.
+    int holds_gil = _PyEval_IsGILEnabled(tstate);
```

Comment:
Does this same pattern exist in `detach_thread()` too? Basically if the GIL is disabled between `tstate_set_detached` and `_PyEval_ReleaseLock`.

I think that instead of checking whether the GIL is enabled, we may want to keep track of whether the current thread holds the GIL (in `PyThreadState` or `_PyThreadStateImpl`). I think we could then push the check down to `drop_gil`, which seems more robust (`PyEval_ReleaseLock` is still sort of publicly exposed).

Reply:
> Does this same pattern exist in `detach_thread()` too?

Yeah (although it should only be possible for another thread to enable the GIL, not disable it). I had convinced myself that it wasn't an issue there but looking at it now, that's obviously not the case.

> I think that instead of checking whether the GIL is enabled, we may want to keep track of whether the current thread holds the GIL

Makes sense; I'll do that.